Index: hibernate-325ga/src/org/hibernate/engine/query/sql/NativeSQLQueryNonScalarReturn.java
===================================================================
--- hibernate-325ga.orig/src/org/hibernate/engine/query/sql/NativeSQLQueryNonScalarReturn.java	2009-08-25 11:54:57.490563801 +0200
+++ hibernate-325ga/src/org/hibernate/engine/query/sql/NativeSQLQueryNonScalarReturn.java	2009-08-25 11:55:45.974977697 +0200
@@ -19,6 +19,7 @@
 	private final String alias;
 	private final LockMode lockMode;
 	private final Map propertyResults = new HashMap();
+	private final int hashCode;
 
 	/**
 	 * Constructs some form of non-scalar return descriptor
@@ -36,6 +37,17 @@
 		if ( propertyResults != null ) {
 			this.propertyResults.putAll( propertyResults );
 		}
+
+		// pre-determine and cache the hashcode
+		int hashCode = alias.hashCode();
+		
+		hashCode = 29 * hashCode + lockMode.toString().hashCode();
+		
+		if ( propertyResults != null ) {
+			hashCode = 29 * hashCode + propertyResults.hashCode();
+		}
+		
+		this.hashCode = hashCode;
 	}
 
 	/**
@@ -64,4 +76,25 @@
 	public Map getPropertyResultsMap() {
 		return Collections.unmodifiableMap( propertyResults );
 	}
+	
+	
+	public boolean equals(Object o) {
+		if ( this == o ) {
+			return true;
+		}
+		if ( o == null || getClass() != o.getClass() ) {
+			return false;
+		}
+
+		final NativeSQLQueryNonScalarReturn that = ( NativeSQLQueryNonScalarReturn ) o;
+		
+		return this.alias.equals(that.alias) 
+			&& this.lockMode.toString().equals(that.lockMode.toString())
+			&& ((this.propertyResults == null && that.propertyResults == null) || (this.propertyResults != null && this.propertyResults.equals(that.propertyResults))); 
+	}
+
+
+	public int hashCode() {
+		return hashCode;
+	}
 }
Index: hibernate-325ga/src/org/hibernate/engine/query/sql/NativeSQLQueryScalarReturn.java
===================================================================
--- hibernate-325ga.orig/src/org/hibernate/engine/query/sql/NativeSQLQueryScalarReturn.java	2009-08-25 11:55:03.358562896 +0200
+++ hibernate-325ga/src/org/hibernate/engine/query/sql/NativeSQLQueryScalarReturn.java	2009-08-25 12:11:27.428310468 +0200
@@ -10,10 +10,20 @@
 public class NativeSQLQueryScalarReturn implements NativeSQLQueryReturn {
 	private Type type;
 	private String columnAlias;
+	private final int hashCode;
 
 	public NativeSQLQueryScalarReturn(String alias, Type type) {
 		this.type = type;
 		this.columnAlias = alias;
+		
+		// pre-determine and cache the hashcode
+		int hashCode = (type == null) ? 0 : type.getClass().getName().hashCode();
+		
+		if (columnAlias != null) {
+			hashCode = hashCode = 29 * hashCode + columnAlias.hashCode();
+		}
+		
+		this.hashCode = hashCode;
 	}
 
 	public String getColumnAlias() {
@@ -23,5 +33,23 @@
 	public Type getType() {
 		return type;
 	}
+	
+	public boolean equals(Object o) {
+		if ( this == o ) {
+			return true;
+		}
+		if ( o == null || getClass() != o.getClass() ) {
+			return false;
+		}
 
+		final NativeSQLQueryScalarReturn that = ( NativeSQLQueryScalarReturn ) o;
+		
+		return ((this.type == null && that.type == null) || this.type.getClass() == that.type.getClass())
+			&& ((this.columnAlias == null && that.columnAlias == null) || (this.columnAlias != null && this.columnAlias.equals(that.columnAlias))); 
+	}
+
+
+	public int hashCode() {
+		return hashCode;
+	}
 }
